consensus/istanbul/backend/internal/enodes/val_enode_db.go

```diff
@@ -256,19 +271,35 @@ func (vet *ValidatorEnodeDB) GetHighestKnownVersionFromAddress(address common.Ad
 	return entry.HighestKnownVersion, nil
 }
 
-// GetAllValEnodes will return all entries in the valEnodeDB
-func (vet *ValidatorEnodeDB) GetAllValEnodes() (map[common.Address]*AddressEntry, error) {
+// GetValEnodes will return entries in the valEnodeDB filtered on the valAddresses parameter.
+// If it's set to nil, then no filter will be applied.
+func (vet *ValidatorEnodeDB) GetValEnodes(valAddresses []common.Address) (map[common.Address]*AddressEntry, error) {
 	vet.lock.RLock()
 	defer vet.lock.RUnlock()
 	var entries = make(map[common.Address]*AddressEntry)
+	var valAddressesMap map[common.Address]struct{}
+
+	if valAddresses != nil {
+	   valAddressesMap := make(map[common.Address]struct{})
+	   for _, address := range valAddresses {
+	       valAddressesMap[address] = struct{}{}
+	   }
+	}
 
 	err := vet.iterateOverAddressEntries(func(address common.Address, entry *AddressEntry) error {
-		entries[address] = entry
+	        if valAddressesMap != nil {
+		   if _, ok := valAddressesMap[address]; ok {
+		      entries[address ] = entry
+		   }
+		} else {
```

Comment:
tip/suggestion.

Instead of checking the condition on every entry, you can do:

```go
var collectorFn func(address common.Address, entry *AddressEntry) error
if valAddresses == nil {
  collectorFn = func(address common.Address, entry *AddressEntry) error {
    entries[address] = entry
  }
} else {
  // build valAddressMap
  collectorFn = func(address common.Address, entry *AddressEntry) error {
    if _, ok := valAddressesMap[address]; ok {
		  entries[address ] = entry
		}
  }
}

err := vet.iterateOverAddressEntries(collectorFn)
```

Reply:
Do you feel strongly about this?  IMO, readability of both options is about equal.

Reply:
Not strongly, just personal taste. So I'm fine if you prefer it the other way.